src/engine/Service.cpp

```diff
@@ -92,6 +95,44 @@ ResultTable Service::computeResult() {
   serviceIriString.remove_suffix(1);
   ad_utility::httpUtils::Url serviceUrl{serviceIriString};
 
+  if (siblingTree_ != nullptr) {
+    // Get the result of the siblingTree, to (potentially)
+    // reduce complexity of the SERVICE query.
+    auto siblingResult = siblingTree_->getResult();
+
+    const size_t rowLimit = 100;
+    if (siblingResult->size() < rowLimit) {
+      auto siblingVariables = siblingTree_->getVariableColumns();
+
+      // Build value clause for each common variable.
+      std::string valueClauses = "{ ";
+      for (const auto& lVar : parsedServiceClause_.visibleVariables_) {
+        auto it = siblingVariables.find(lVar);
+        if (it == siblingVariables.end()) {
```

Comment:
First in a separate step compute the relevant column indices for the variables, and then simply iterate over those.
That makes the large loop more readable and possibly also more efficient.